### ISSUE TITLE
[Discussion] Service Bus - Track 2 - Management client model changes

### ISSUE THREAD

**Issue by JoshLove-msft** [Azure/azure-sdk-for-net#13352]
_The intent of this issue is to solicit feedback from the community in advance of the GA of the next major version of the Service Bus library (which will follow several Preview releases).
As a reference, the .NET Azure SDK guidelines can be found here: https://azure.github.io/azure-sdk/dotnet_introduction.html
Terminology:
Track 0 refers to the [WindowsAzure.ServiceBus library](https://www.nuget.org/packages/WindowsAzure.ServiceBus/)
Track 1 refers to the [Microsoft.Azure.ServiceBus library](https://www.nuget.org/packages/Microsoft.Azure.ServiceBus/)
Track 2 refers to the new library, [Azure.Messaging.ServiceBus](https://www.nuget.org/packages/Azure.Messaging.ServiceBus/), we are working on releasing._

In the Track 1 [ManagementClient](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Microsoft.Azure.ServiceBus/src/Management/ManagementClient.cs), there is a single model type, `QueueDescription`, that is used for both creating and updating queues. A similar pattern exists for topics, subscriptions, and rules. Since this same type is used for both creating and updating queues, all of the properties are settable even though 4 of the 15 properties on a queue cannot actually be changed after the queue is created. If a user attempts to change one of those properties, they will hit a runtime error. 

In order to make these restrictions more evident in the API itself, we are proposing using two separate types to model each entity - one for creation, and one for updating. This new type will be named as `Create(Queue|Topic|Rule|Subscription)Options`. We are also proposing renaming `(Queue|Topic|Rule|Subscription)Description` to be `(Queue|Topic|Rule|Subscription)Properties`, as this is more idiomatic in the Azure SDK Track 2 landscape. In order to maintain consistency with the  `(Queue|Topic|Rule|Subscription)Properties` name, we also propose renaming `(Queue|Topic|Subscription)RuntimeInfo` to `(Queue|Topic|Subscription)RuntimeProperties`.

Creating a queue:
```c#
var options = new CreateQueueOptions(queueName)
{
    AutoDeleteOnIdle = TimeSpan.FromHours(1),
    DefaultMessageTimeToLive = TimeSpan.FromDays(2),
    DuplicateDetectionHistoryTimeWindow = TimeSpan.FromMinutes(1),
    EnableBatchedOperations = true,
    DeadLetteringOnMessageExpiration = true,
    EnablePartitioning = false,
    ForwardDeadLetteredMessagesTo = null,
    ForwardTo = null,
    LockDuration = TimeSpan.FromSeconds(45),
    MaxDeliveryCount = 8,
    MaxSizeInMegabytes = 2048,
    RequiresDuplicateDetection = true,
    RequiresSession = true,
    UserMetadata = nameof(BasicQueueCrudOperations)
};
QueueProperties queue = await client.CreateQueueAsync(options);
```

In order to update a queue, you would pass in the `QueueProperties` after making your changes. Only properties that can be changed on an already existing queue will have setters.
```c#
QueueProperties queue = await client.GetQueueAsync(queueName);
queue.MaxDeliveryCount = 5;
QueueProperties updated = await client.UpdateQueue(queue);
```

In order to support creating a new queue based on an existing queue, we will include a constructor for `CreateQueueOptions` that takes a `QueueProperties`:
```c#
QueueProperties queue = await client.GetQueueAsync(queueName);
var options = new CreateQueueOptions(queue);
options.Name = "new-queue";
QueueProperties newQueue = await client.CreateQueueAsync(options);
```

And to look up runtime properties:
```c#
QueueRuntimeProperties runtime = await client.GetQueueRuntimePropertiesAsync(newQueue.Name);
```

**Comment by SeanFeldman** [comment]
1. For the get properties methods on the management client, e.g. `QueueProperties`, I would suggest to have a method that is less ambiguous than `GetQueueAsync(queueName)`. A `GetQueuePropertiesAsync(queueName)` would be a better for intent revealing.

1. While I see the rationale behind the naming chosen for `CreateQueueOptions` and `QueueProperties`, I find it confusing to have "options" and "properties" while both stand for the same things. `CreateQueueProperties` and `QueueProperties` would resonate better with me.

I'm assuming along with `QueueDescription` there will be the analogous constructs for subscriptions, topics, and rules?

**Comment by JoshLove-msft** [comment]
> 1. For the get properties methods on the management client, e.g. `QueueProperties`, I would suggest to have a method that is less ambiguous than `GetQueueAsync(queueName)`. A `GetQueuePropertiesAsync(queueName)` would be a better for intent revealing.
> 2. While I see the rationale behind the naming chosen for `CreateQueueOptions` and `QueueProperties`, I find it confusing to have "options" and "properties" while both stand for the same things. `CreateQueueProperties` and `QueueProperties` would resonate better with me.
> 
> I'm assuming along with `QueueDescription` there will be the analogous constructs for subscriptions, topics, and rules?

1. `GetQueuePropertiesAsync` seems reasonable on its own, but would it be strange that we would still have a `CreateQueueAsync` method that returns `QueueProperties`?

2. I can see your point about having two distinct but similar suffixes here potentially adding to the cognitive burden, but we do try to use the options pattern across track 2 for input property bags containing optional properties, so hopefully it is a pattern that users will become accustomed to.

Yes, we will have analogous changes for the other entities.

**Comment by SeanFeldman** [comment]
> 1. `GetQueuePropertiesAsync` seems reasonable on its own, but would it be strange that we would still have a `CreateQueueAsync` method that returns `QueueProperties`?

`GetQueuePropertiesAsync` is a method to return queue properties. For the scenarios where a user wants to get queue properties, it will be natural to look for this method to invoke. `GetQueue` feels off as the name implies that you get a **queue** or a queue reference (see older Storage SDKs).

As for `CreateQueueAsync`, looking from a command-query separation perspective, yes, it's a bit odd. If the broker can return successfully and not apply all the properties, it would be a must to have the return value. As far as I know, that's not the case and the broker would throw if it cannot construct an entity with a given entity description (properties/create options). Ironically, I'll use your own design against your point here 😈: the method is provided with `CreateQueueOptions` but returns `QueueProperties`

**Comment by SeanFeldman** [comment]
> Yes, we will have analogous changes for the other entities.

👍 
You've got that in the issue description. Thank you.

**Comment by JoshLove-msft** [comment]
> `GetQueuePropertiesAsync` is a method to return queue properties. For the scenarios where a user wants to get queue properties, it will be natural to look for this method to invoke. `GetQueue` feels off as the name implies that you get a **queue** or a queue reference (see older Storage SDKs).

I would tend to agree that on its own `GetQueuePropertiesAsync` is preferable. The part I am less sure of is whether the tradeoff of changing this name is worth the slight inconsistency with the naming of and `CreateQueueAsync`.  As you mention though, the input type is already different for `CreateQueueAsync`, so maybe it wouldn't be that jarring for users.

The other piece that I hadn't thought of initially is what this would to the [API](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/servicebus/Azure.Messaging.ServiceBus/src/Management/ServiceBusManagementClient.cs#L516) that gets all queues in the namespace - would this be `GetQueuesPropertiesAsync`? Despite being an accurate name, it seems hard to read and less intuitive than the simple `GetQueuesAsync`. In the end, this will be a tradeoff one way or the other.

**Comment by SeanFeldman** [comment]
Personally, I'd take `GetQueuesPropertiesAsync` over `GetQueuesAsync`.

**Comment by JoshLove-msft** [comment]
> Personally, I'd take `GetQueuesPropertiesAsync` over `GetQueuesAsync`.

Will talk this over with feature crew and PG. Here is my summary of the tradeoffs:

**GetQueueAsync/UpdateQueueAsync**
 - consistent with CreateQueueAsync which is nice considering the return types are all the same.
 - not quite as accurate/descriptive as GetQueuePropertiesAsync, though the return type should hopefully clear up any confusion

**GetQueuePropertiesAsync/UpdateQueuePropertiesAsync**
- would be inconsistent with CreateQueueAsync.
- more descriptive and technically accurate
- Would be forced into using the awkward name `GetQueuesPropertiesAsync` for the get all queues API.

**Comment by JoshLove-msft** [comment]
I updated the issue description to mention that actually only 4 of the queue properties cannot be modified on created queues - EnablePartitioning, RequiresSession, RequiresDuplicateDetection, and Name. This arguably makes this change less of a slam dunk, but I still think it is an improvement and makes for a more intuitive API. Splitting into two types and requiring a get before an update also allows us to make the partial update semantics less ambiguous - previously a user would be able to new up a `QueueDescription` directly and attempt to make an update. Doing this would result in resetting all of the properties that they left blank to their default values, which may not be what users expect. With two separate types, we have restricted the ability to new up `QueueProperties`, so the user is forced to get the queue before doing an update.
